File: src/rules/game_rules.py
# Game Rules System - Core game mechanics and interactions

from typing import List, Dict, Any, Optional, Tuple
from src.world.game_world import GameWorld
from src.entities.base import GameObject
from src.entities.objects.flame import Flame
from src.entities.objects.ice_block import IceBlock
from src.entities.objects.stone import Stone
from src.entities.objects.pot import Pot
from src.physics.physics_engine import PhysicsEngine
from src.physics.ice_system import IceBlockSystem


class GameRulesSystem:
    """System for managing core game rules and win conditions"""
    
    def __init__(self, game_world: GameWorld, physics_engine: PhysicsEngine, ice_system: IceBlockSystem):
        self.game_world = game_world
        self.physics_engine = physics_engine
        self.ice_system = ice_system
        
        # Game state
        self.is_level_complete = False
        self.game_won = False
        self.moves_taken = 0
        self.time_elapsed = 0.0
        self.puzzles_solved = []
        self.player = None
        self.player = None
        
        # Rule configuration
        self.max_moves = 50  # Maximum moves before level fails (optional)
        self.time_limit = 300  # 5 minutes time limit (optional)
        
        # Win conditions
        self.require_all_flames_extinguished = True
        self.require_player_at_exit = False  # For specific levels
        self.require_min_objects_collected = 0  # Optional collectible requirement
        
        # Lose conditions
        self.lose_on_player_death = True
        self.lose_on_time_limit = False
        self.lose_on_move_limit = False
    
    def update(self, dt: float):
        """Update game rules"""
        self.time_elapsed += dt
        
        # Check win conditions
        self._check_win_conditions()
        
        # Check lose conditions
        self._check_lose_conditions()
        
        # Update game-specific rules
        self._update_game_specific_rules(dt)
    
    def _check_win_conditions(self):
        """Check if level is won"""
        if self.is_level_complete:
            return
        
        # Primary win condition: All flames extinguished
        if self.require_all_flames_extinguished:
            flame_count = self.game_world.count_objects_of_type(Flame)
            if flame_count == 0:
                self._trigger_level_complete()
                return
        
        # Secondary conditions
        if self.require_player_at_exit:
            if self._check_player_at_exit():
                self._trigger_level_complete()
                return
        
        if self.require_min_objects_collected > 0:
            collected = len(self.puzzles_solved)
            if collected >= self.require_min_objects_collected:
                self._trigger_level_complete()
                return
    
    def _check_lose_conditions(self):
        """Check if level is lost"""
        # Check time limit
        if self.lose_on_time_limit and self.time_elapsed > self.time_limit:
            self._trigger_level_failed("Time limit exceeded")
            return
        
        # Check move limit
        if self.lose_on_move_limit and self.moves_taken > self.max_moves:
            self._trigger_level_failed("Move limit exceeded")
            return
        
        # Check player death
        if self.lose_on_player_death:
            if self._check_player_death():
                self._trigger_level_failed("Player died")
                return
    
    def _update_game_specific_rules(self, dt: float):
        """Update game-specific rules"""
        # Flame extinguishing by ice
        self._process_flame_ice_interactions()
        
        # Ice pot ignition by flame
        self._process_pot_flame_interactions()
        
        # Ice block melting by hot pot
        self._process_ice_hotpot_interactions()
        
        # Stone heat resistance
        self._process_stone_heat_interactions()
        
        # Portal transportation
        self._process_portal_interactions()
        
        # Ice chain reactions
        self._process_ice_chain_reactions()
    
    def _process_flame_ice_interactions(self):
        """Process flame and ice block interactions"""
        flames = self.game_world.find_objects_of_type(Flame)
        ice_blocks = self.game_world.find_objects_of_type(IceBlock)
        
        for fx, fy, flame in flames:
            for ix, iy, ice in ice_blocks:
                # Check if flame and ice are adjacent
                if abs(fx - ix) <= 1 and abs(fy - iy) <= 1:
                    # Check if ice can extinguish flame
                    if self._can_ice_extinguish_flame(ice, flame):
                        self._extinguish_flame_with_ice(flame, ice)
    
    def _can_ice_extinguish_flame(self, ice, flame) -> bool:
        """Check if ice can extinguish flame"""
        # Ice can extinguish flame if they're in contact
        # Note: Ice above flame doesn't melt (per rules)
        return True
    
    def _extinguish_flame_with_ice(self, flame: Flame, ice: IceBlock):
        """Handle flame extinguishing by ice"""
        if flame.is_burning():
            flame.extinguish()
            self.game_world.remove_object(flame)
            self.ice_system.remove_ice_block(ice, self.game_world)
            
            # Add to solved puzzles
            self.puzzles_solved.append(f"Extinguished flame at ({flame.grid_x},{flame.grid_y})")
    
    def _process_pot_flame_interactions(self):
        """Process pot and flame interactions"""
        pots = self.game_world.find_objects_of_type(Pot)
        flames = self.game_world.find_objects_of_type(Flame)
        
        for px, py, pot in pots:
            if pot.is_ice():
                for fx, fy, flame in flames:
                    # Check if flame is adjacent to ice pot
                    if abs(px - fx) <= 1 and abs(py - fy) <= 1:
                        # Check if pot can be ignited
                        if self._can_flame_ignite_pot(pot, flame):
                            pot.check_for_ignition(self.game_world)
                            if pot.is_hot():
                                self.puzzles_solved.append(f"Ignited pot at ({pot.grid_x},{pot.grid_y})")
    
    def _can_flame_ignite_pot(self, pot: Pot, flame: Flame) -> bool:
        """Check if flame can ignite pot"""
        return pot.is_ice() and pot.can_ignite() and flame.is_burning()
    
    def _process_ice_hotpot_interactions(self):
        """Process ice block and hot pot interactions"""
        ice_blocks = self.game_world.find_objects_of_type(IceBlock)
        pots = self.game_world.find_objects_of_type(Pot)
        
        for ix, iy, ice in ice_blocks:
            for px, py, pot in pots:
                if pot.is_hot():
                    # Check if ice is directly above hot pot
                    if ix == px and iy == py + 1:
                        # Check if ice can be melted
                        if self._can_hotpot_melt_ice(pot, ice):
                            self._melt_ice_with_hotpot(ice, pot)
    
    def _can_hotpot_melt_ice(self, pot: Pot, ice: IceBlock) -> bool:
        """Check if hot pot can melt ice"""
        return pot.is_hot() and ice.can_be_placed_on_hot_pot(pot.grid_y)
    
    def _melt_ice_with_hotpot(self, ice: IceBlock, pot: Pot):
        """Handle ice melting by hot pot"""
        if ice.is_active:
            self.ice_system.remove_ice_block(ice, self.game_world)
            self.puzzles_solved.append(f"Melted ice at ({ice.grid_x},{ice.grid_y}) with hot pot")
    
    def _process_stone_heat_interactions(self):
        """Process stone heat resistance"""
        stones = self.game_world.find_objects_of_type(Stone)
        flames = self.game_world.find_objects_of_type(Flame)
        pots = self.game_world.find_objects_of_type(Pot)
        
        # Stones are heat resistant, so they don't get destroyed by flames or hot pots
        # This is more about protecting stones from being removed by interactions
        pass
    
    def _process_portal_interactions(self):
        """Process portal transportation rules"""
        from src.entities.objects.portal import Portal
        
        portals = self.game_world.find_objects_of_type(Portal)
        player = self._get_player()
        
        if not player:
            return
        
        # Check if player is at portal position
        for x, y, portal in portals:
            if x == player.grid_x and y == player.grid_y:
                # Check if player can use portal
                if portal.can_player_enter(player.grid_x, player.grid_y):
                    # Check portal height difference
                    height_diff = portal.grid_y - player.grid_y
                    max_diff = portal.get_property('height_difference', 1)
                    
                    if height_diff <= max_diff:
                        # Transport player
                        linked_portal = portal.get_linked_portal()
                        if linked_portal and linked_portal.can_player_accept_player(player, self.game_world):
                            portal.transport_player(player, self.game_world)
                            self.puzzles_solved.append(f"Used portal from ({portal.grid_x},{portal.grid_y})")
    
    def _process_ice_chain_reactions(self):
        """Process ice chain reactions"""
        chains = self.ice_system.get_ice_chain_reactions(self.game_world)
        
        for chain in chains:
            if len(chain) > 1:
                # Record chain reaction as solved puzzle
                start_pos = f"({chain[0].grid_x},{chain[0].grid_y})"
                self.puzzles_solved.append(f"Triggered chain reaction starting at {start_pos}")
    
    def _check_player_at_exit(self) -> bool:
        """Check if player is at exit position"""
        player = self._get_player()
        if not player:
            return False
        
        # TODO: Implement exit position checking
        return False
    
    def _check_player_death(self) -> bool:
        """Check if player is dead"""
        player = self._get_player()
        if not player:
            return False
        
        # TODO: Implement player death checking
        return False
    
    def _get_player(self) -> Optional[GameObject]:
        """Get player object"""
        # Find player in world
        for x, y, obj in self.game_world.grid:
            if obj and obj.get_type() == "player":
                return obj
        return None
    
    def _trigger_level_complete(self):
        """Trigger level completion"""
        self.is_level_complete = True
        self.game_won = True
        print(f"🎉 Level Complete! Time: {self.time_elapsed:.1f}s, Moves: {self.moves_taken}")
        print(f"Puzzles solved: {len(self.puzzles_solved)}")
        for puzzle in self.puzzles_solved:
            print(f"  • {puzzle}")
    
    def _trigger_level_failed(self, reason: str):
        """Trigger level failure"""
        self.game_won = False
        print(f"❌ Level Failed: {reason}")
        print(f"Time: {self.time_elapsed:.1f}s, Moves: {self.moves_taken}")
    
    def reset_level(self):
        """Reset level state"""
        self.is_level_complete = False
        self.game_won = False
        self.moves_taken = 0
        self.time_elapsed = 0.0
        self.puzzles_solved.clear()
    
    def increment_moves(self):
        """Increment move counter"""
        self.moves_taken += 1
    
    def get_level_status(self) -> Dict[str, Any]:
        """Get current level status"""
        return {
            'is_complete': self.is_level_complete,
            'game_won': self.game_won,
            'moves_taken': self.moves_taken,
            'time_elapsed': self.time_elapsed,
            'flames_remaining': self.game_world.count_objects_of_type(Flame),
            'ice_blocks_count': self.game_world.count_objects_of_type(IceBlock),
            'puzzles_solved': len(self.puzzles_solved),
            'puzzles_list': self.puzzles_solved.copy()
        }
    
    def set_level_requirements(self, **requirements):
        """Set level requirements"""
        if 'require_all_flames_extinguished' in requirements:
            self.require_all_flames_extinguished = requirements['require_all_flames_extinguished']
        
        if 'require_player_at_exit' in requirements:
            self.require_player_at_exit = requirements['require_player_at_exit']
        
        if 'require_min_objects_collected' in requirements:
            self.require_min_objects_collected = requirements['require_min_objects_collected']
        
        if 'time_limit' in requirements:
            self.time_limit = requirements['time_limit']
            self.lose_on_time_limit = True
        
        if 'move_limit' in requirements:
            self.max_moves = requirements['move_limit']
            self.lose_on_move_limit = True
    
    def add_puzzle_solved(self, puzzle: str):
        """Add a solved puzzle"""
        if puzzle not in self.puzzles_solved:
            self.puzzles_solved.append(puzzle)
    
    def is_game_over(self) -> bool:
        """Check if game is over (won or lost)"""
        return self.is_level_complete

File: src/rules/gameRules.ts
import { GameWorld } from '@/world/gameWorld';
import { PhysicsEngine } from './physicsEngine';
import { IceBlockSystem } from './iceSystem';
import { GameObject } from '@/entities/base';

/**
 * Game rules system for handling object interactions
 */
export class GameRulesSystem {
  private gameWorld: GameWorld;
  private physicsEngine: PhysicsEngine;
  private iceSystem: IceBlockSystem;

  constructor(gameWorld: GameWorld, physicsEngine: PhysicsEngine, iceSystem: IceBlockSystem) {
    this.gameWorld = gameWorld;
    this.physicsEngine = physicsEngine;
    this.iceSystem = iceSystem;
  }

  /**
   * Update game rules
   */
  update(dt: number): void {
    // Check all object interactions
    this.checkObjectInteractions();
    
    // Handle special game rules
    this.handleSpecialRules();
    
    // Clean up destroyed objects
    this.cleanupDestroyedObjects();
  }

  /**
   * Check interactions between objects
   */
  private checkObjectInteractions(): void {
    const objects = this.gameWorld.getAllObjects();
    
    for (let i = 0; i < objects.length; i++) {
      for (let j = i + 1; j < objects.length; j++) {
        const obj1 = objects[i];
        const obj2 = objects[j];
        
        if (this.areInteracting(obj1, obj2)) {
          this.handleInteraction(obj1, obj2);
        }
      }
    }
  }

  /**
   * Check if two objects are interacting
   */
  private areInteracting(obj1: GameObject, obj2: GameObject): boolean {
    // Objects interact if they're in the same cell or adjacent cells
    const sameCell = obj1.gridX === obj2.gridX && obj1.gridY === obj2.gridY;
    
    if (sameCell) return true;
    
    // Check adjacent cells for certain interactions
    const adjacentX = Math.abs(obj1.gridX - obj2.gridX) === 1 && obj1.gridY === obj2.gridY;
    const adjacentY = Math.abs(obj1.gridY - obj2.gridY) === 1 && obj1.gridX === obj2.gridX;
    
    return adjacentX || adjacentY;
  }

  /**
   * Handle interaction between two objects
   */
  private handleInteraction(obj1: GameObject, obj2: GameObject): void {
    const type1 = obj1.getType();
    const type2 = obj2.getType();

    // Fire interactions
    this.handleFireInteractions(obj1, obj2);
    
    // Ice interactions
    this.handleIceInteractions(obj1, obj2);
    
    // Pot interactions
    this.handlePotInteractions(obj1, obj2);
    
    // Portal interactions
    this.handlePortalInteractions(obj1, obj2);
    
    // Stone interactions
    this.handleStoneInteractions(obj1, obj2);
  }

  /**
   * Handle fire-related interactions
   */
  private handleFireInteractions(obj1: GameObject, obj2: GameObject): void {
    if (obj1.getType() === 'flame' || obj2.getType() === 'flame') {
      const flame = obj1.getType() === 'flame' ? obj1 : obj2;
      const other = obj1.getType() === 'flame' ? obj2 : obj1;
      
      // Fire ignites cold pots
      if (other.getType() === 'pot' && other.getProperty('is_cold')) {
        this.ignitePot(other, flame);
      }
      
      // Fire melts ice blocks
      if (other.getType() === 'ice_block') {
        this.meltIceBlock(other, flame);
      }
      
      // Fire spreads to flammable objects
      if (other.getProperty('flammable')) {
        this.spreadFire(flame, other);
      }
    }
  }

  /**
   * Handle ice-related interactions
   */
  private handleIceInteractions(obj1: GameObject, obj2: GameObject): void {
    if (obj1.getType() === 'ice_block' || obj2.getType() === 'ice_block') {
      const ice = obj1.getType() === 'ice_block' ? obj1 : obj2;
      const other = obj1.getType() === 'ice_block' ? obj2 : obj1;
      
      // Ice extinguishes flames
      if (other.getType() === 'flame') {
        this.extinguishFlame(other, ice);
      }
      
      // Ice cools hot pots
      if (other.getType() === 'pot' && other.getProperty('is_hot')) {
        this.coolPot(other, ice);
      }
    }
  }

  /**
   * Handle pot interactions
   */
  private handlePotInteractions(obj1: GameObject, obj2: GameObject): void {
    if (obj1.getType() === 'pot' || obj2.getType() === 'pot') {
      const pot1 = obj1.getType() === 'pot' ? obj1 : obj2;
      const pot2 = obj1.getType() === 'pot' ? obj2 : obj1;
      
      // Pot to pot interactions (temperature exchange)
      if (pot2.getType() === 'pot') {
        this.handlePotPotInteraction(pot1, pot2);
      }
    }
  }

  /**
   * Handle portal interactions
   */
  private handlePortalInteractions(obj1: GameObject, obj2: GameObject): void {
    if (obj1.getType() === 'portal' || obj2.getType() === 'portal') {
      const portal = obj1.getType() === 'portal' ? obj1 : obj2;
      const other = obj1.getType() === 'portal' ? obj2 : obj1;
      
      // Portal teleports compatible objects
      if (this.canTeleport(other)) {
        (portal as any).onCollision?.(other);
      }
    }
  }

  /**
   * Handle stone interactions
   */
  private handleStoneInteractions(obj1: GameObject, obj2: GameObject): void {
    if (obj1.getType() === 'stone' || obj2.getType() === 'stone') {
      const stone = obj1.getType() === 'stone' ? obj1 : obj2;
      const other = obj1.getType() === 'stone' ? obj2 : obj1;
      
      // Stone crushes fragile objects when falling
      if (this.isFalling(stone) && other.isFragile()) {
        this.crushObject(other, stone);
      }
    }
  }

  /**
   * Handle special game rules
   */
  private handleSpecialRules(): void {
    // Check win conditions
    this.checkWinCondition();
    
    // Check lose conditions
    this.checkLoseCondition();
    
    // Apply environmental effects
    this.applyEnvironmentalEffects();
  }

  /**
   * Check win condition
   */
  private checkWinCondition(): void {
    const flameCount = this.gameWorld.countObjectsOfType('flame');
    // Win condition handled in game class
  }

  /**
   * Check lose condition
   */
  private checkLoseCondition(): void {
    // Check if player is destroyed
    const player = this.gameWorld.getObjectsOfType('player')[0];
    if (player && !player.isActive) {
      // Player lost - this would be handled in game class
    }
  }

  /**
   * Apply environmental effects
   */
  private applyEnvironmentalEffects(): void {
    // Melt ice blocks near heat sources
    this.applyHeatEffects();
    
    // Extinguish flames near water/ice
    this.applyCoolingEffects();
  }

  /**
   * Apply heat effects to nearby objects
   */
  private applyHeatEffects(): void {
    const iceBlocks = this.gameWorld.getObjectsOfType('ice_block');
    const hotObjects = this.gameWorld.getAllObjects().filter(obj => 
      obj.getType() === 'flame' || 
      (obj.getType() === 'pot' && obj.getProperty('is_hot'))
    );

    for (const iceBlock of iceBlocks) {
      for (const hotObject of hotObjects) {
        const distance = Math.sqrt(
          Math.pow(iceBlock.gridX - hotObject.gridX, 2) + 
          Math.pow(iceBlock.gridY - hotObject.gridY, 2)
        );
        
        if (distance <= 2) { // Heat affects objects within 2 cells
          (iceBlock as any).startMelting?.();
        }
      }
    }
  }

  /**
   * Apply cooling effects to nearby objects
   */
  private applyCoolingEffects(): void {
    const flames = this.gameWorld.getObjectsOfType('flame');
    const coldObjects = this.gameWorld.getAllObjects().filter(obj => 
      obj.getType() === 'ice_block' || 
      (obj.getType() === 'pot' && obj.getProperty('is_cold'))
    );

    for (const flame of flames) {
      for (const coldObject of coldObjects) {
        const distance = Math.sqrt(
          Math.pow(flame.gridX - coldObject.gridX, 2) + 
          Math.pow(flame.gridY - coldObject.gridY, 2)
        );
        
        if (distance <= 1) { // Cooling affects objects within 1 cell
          if (coldObject.getType() === 'ice_block') {
            // Ice block extinguishes flame
            this.extinguishFlame(flame, coldObject);
          }
        }
      }
    }
  }

  /**
   * Clean up destroyed objects
   */
  private cleanupDestroyedObjects(): void {
    const objects = this.gameWorld.getAllObjects();
    
    for (const object of objects) {
      if (!object.isActive) {
        this.gameWorld.removeObject(object.gridX, object.gridY);
      }
    }
  }

  // Helper methods for specific interactions
  private ignitePot(pot: GameObject, flame: GameObject): void {
    (pot as any).startHeating?.();
    flame.destroy();
  }

  private meltIceBlock(ice: GameObject, flame: GameObject): void {
    (ice as any).startMelting?.();
  }

  private extinguishFlame(flame: GameObject, ice: GameObject): void {
    flame.destroy();
    (ice as any).startMelting?.();
  }

  private coolPot(pot: GameObject, ice: GameObject): void {
    (pot as any).coolDown?.();
    ice.destroy();
  }

  private handlePotPotInteraction(pot1: GameObject, pot2: GameObject): void {
    const pot1Hot = pot1.getProperty('is_hot');
    const pot2Hot = pot2.getProperty('is_hot');
    
    // Temperature exchange
    if (pot1Hot && !pot2Hot) {
      (pot1 as any).coolDown?.();
      (pot2 as any).startHeating?.();
    } else if (!pot1Hot && pot2Hot) {
      (pot2 as any).coolDown?.();
      (pot1 as any).startHeating?.();
    }
  }

  private canTeleport(obj: GameObject): boolean {
    const teleportableTypes = ['player', 'ice_block', 'stone', 'pot'];
    return teleportableTypes.includes(obj.getType());
  }

  private isFalling(obj: GameObject): boolean {
    // Check if object is falling (simplified check)
    return obj.getProperty('falling', false);
  }

  private crushObject(obj: GameObject, crusher: GameObject): void {
    if (crusher.getWeight() > obj.getWeight()) {
      obj.destroy();
    }
  }

  private spreadFire(flame: GameObject, target: GameObject): void {
    if (target.getProperty('flammable')) {
      // Create new flame at target position
      // This would create a new flame object
      target.destroy();
    }
  }
}